File: examples/as-fs/src/console.ts
// Console API local copy for AS examples

@external("wasi_snapshot_preview1", "fd_write")
declare function wasiFdWrite(fd: i32, iovs: usize, iovsLen: i32, nwritten: usize): i32;

export function log(msg: string): void {
    let buf = String.UTF8.encode(msg + "\n");
    let arr = Uint8Array.wrap(buf);
    let iovsAddr = heap.alloc(8);
    store<usize>(iovsAddr, arr.dataStart as usize, 0);
    store<i32>(iovsAddr, arr.length, 4);
    store<i32>(iovsAddr, 0, 8);
    let nwritten = heap.alloc(4);
    wasiFdWrite(1, iovsAddr, 1, nwritten);
}

export function warn(msg: string): void {
    let buf = String.UTF8.encode(msg + "\n");
    let arr = Uint8Array.wrap(buf);
    let iovsAddr = heap.alloc(8);
    store<usize>(iovsAddr, arr.dataStart as usize, 0);
    store<i32>(iovsAddr, arr.length, 4);
    store<i32>(iovsAddr, 0, 8);
    let nwritten = heap.alloc(4);
    wasiFdWrite(2, iovsAddr, 1, nwritten);
}

export function error(msg: string): void {
    let buf = String.UTF8.encode(msg + "\n");
    let arr = Uint8Array.wrap(buf);
    let iovsAddr = heap.alloc(8);
    store<usize>(iovsAddr, arr.dataStart as usize, 0);
    store<i32>(iovsAddr, arr.length, 4);
    store<i32>(iovsAddr, 0, 8);
    let nwritten = heap.alloc(4);
    wasiFdWrite(2, iovsAddr, 1, nwritten);
}

export function assert(cond: bool, msg: string): void {
    if (!cond) {
        let full = "Assertion failed: " + msg + "\n";
        let buf = String.UTF8.encode(full);
        let arr = Uint8Array.wrap(buf);
        let iovsAddr = heap.alloc(8);
        store<usize>(iovsAddr, arr.dataStart as usize, 0);
        store<i32>(iovsAddr, arr.length, 4);
        store<i32>(iovsAddr, 0, 8);
        let nwritten = heap.alloc(4);
        wasiFdWrite(2, iovsAddr, 1, nwritten);
        unreachable();
    }
}

File: examples/rust-fs/src/console.rs
// Console API local copy for examples
#![allow(non_snake_case, non_camel_case_types, dead_code)]

use crate::wasi;

pub fn log(msg: &str) -> Result<usize, i32> {
    let n1 = wasi::stdout_str(msg).map_err(|e| e.0)?;
    let n2 = wasi::stdout_write(b"\n").map_err(|e| e.0)?;
    Ok(n1 + n2)
}

pub fn warn(msg: &str) -> Result<usize, i32> {
    let n1 = wasi::stderr_write(msg.as_bytes()).map_err(|e| e.0)?;
    let n2 = wasi::stderr_write(b"\n").map_err(|e| e.0)?;
    Ok(n1 + n2)
}

pub fn error(msg: &str) -> Result<usize, i32> {
    let n1 = wasi::stderr_write(msg.as_bytes()).map_err(|e| e.0)?;
    let n2 = wasi::stderr_write(b"\n").map_err(|e| e.0)?;
    Ok(n1 + n2)
}

pub fn assert(cond: bool, msg: &str) -> Result<(), i32> {
    if !cond {
        let _ = wasi::stderr_write(b"Assertion failed: ");
        let _ = wasi::stderr_write(msg.as_bytes());
        let _ = wasi::stderr_write(b"\n");
        return Err(1);
    }
    Ok(())
}
